### PR TITLE
Per-head K/V projection in Physics_Attention_Irregular_Mesh

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

In `Physics_Attention_Irregular_Mesh`, the K/V projection collapses all heads' slice tokens to a shared mean before computing keys and values:

```python
slice_token_kv = slice_token.mean(dim=1, keepdim=True)  # [B, 1, slice_num, d]
K = self.to_k(slice_token_kv)
V = self.to_v(slice_token_kv)
```

This means every attention head queries the **same global mean** key-value pool — there is no head specialization in what is stored in K/V. Head 0 might specialize in pressure gradients, head 1 in wake patterns, head 2 in surface boundary layer — but they all write to the same K/V bank, which the mean actively homogenizes.

**Fix:** Remove the `.mean(dim=1, keepdim=True)` collapse. Each head uses its own slice tokens directly for K/V projection. Q already comes from per-head slice tokens, so this makes Q, K, V all per-head — a symmetric and more natural attention formulation.

**Why it might help p_tan:** The tandem interaction involves long-range wake effects from the fore foil reaching the aft. A head that specializes in inter-foil communication needs its own K/V, not one contaminated by the mean of a boundary-layer-focused head. Removing the shared mean bottleneck allows genuine head specialization.

**Parameter count:** Unchanged — `to_k` and `to_v` are `nn.Linear(d, d_head)` applied per-head. The same number of parameters, just applied to per-head inputs instead of the head-mean. The einsum `'bhnd,bhsd->bhns'` already handles the per-head case.

### Code change in `cfd_tandemfoil/train.py`

In `Physics_Attention_Irregular_Mesh.forward`, find:
```python
slice_token_kv = slice_token.mean(dim=1, keepdim=True)  # collapse to shared mean
K = self.to_k(slice_token_kv)
V = self.to_v(slice_token_kv)
```

Replace with (gated by the new flag):
```python
if self.per_head_kv:
    # Each head uses its own slice tokens for K/V — no mean collapse
    K = self.to_k(slice_token)   # [B, H, slice_num, d_head]
    V = self.to_v(slice_token)   # [B, H, slice_num, d_head]
else:
    slice_token_kv = slice_token.mean(dim=1, keepdim=True)
    K = self.to_k(slice_token_kv)
    V = self.to_v(slice_token_kv)
```

In `Physics_Attention_Irregular_Mesh.__init__`, add:
```python
self.per_head_kv = per_head_kv  # bool, passed from config
```

In `Config`, add:
```python
per_head_kv: bool = False
```

Wire `cfg.per_head_kv` through to the attention module construction.

**IMPORTANT:** Check the einsum shapes. If `to_k` expects `[B, 1, slice_num, d]` and you're now passing `[B, H, slice_num, d]`, it should broadcast correctly since it's an `nn.Linear` applied to the last dim. Verify the attention computation `attn = einsum('bhnd,bhsd->bhns', Q, K)` still works — it should, since both Q and K are now `[B, H, slice_num, d_head]`. The softmax and V aggregation should also be unchanged.

**Training stability:** The shared-mean K/V was implicitly regularizing (all heads see the same keys). Removing it may increase training variance. If loss spikes in the first 20 epochs, add a small initialization delta: zero-initialize `to_k` and `to_v` and add a residual `K = self.to_k(slice_token) + slice_token` for the first run to preserve the effective baseline behavior at init.

### Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent askeladd \
  --wandb_name "askeladd/per-head-kv-s42" \
  --wandb_group "askeladd/per-head-kv-slice" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --per_head_kv

# Seed 73
cd cfd_tandemfoil && python train.py --agent askeladd \
  --wandb_name "askeladd/per-head-kv-s73" \
  --wandb_group "askeladd/per-head-kv-slice" \
  --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --per_head_kv
```

### Report

Table: p_in, p_oodc, p_tan, p_re for both seeds and 2-seed avg vs baseline. W&B run IDs.

**Watch for:** If training loss is unstable in first 30 epochs (spikes), try zero-initializing `to_k`/`to_v` with residual. Report VRAM usage — should be slightly higher since we're no longer collapsing to 1 head before K/V projection.

**Early abort:** If after epoch 50, all four metrics are >5% worse than a naive mid-training baseline, abort and report.

## Baseline (PR #2213, 2-seed avg)

| Metric | Value | Target to beat |
|--------|-------|----------------|
| p_in | 11.979 | < 11.98 |
| p_oodc | 7.643 | < 7.65 |
| p_tan | 28.341 | < 28.34 |
| p_re | 6.300 | < 6.30 |

Baseline W&B runs: `hgml7i2r` (s42), `qic03vrg` (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Implementation

Added `per_head_kv` config flag. When enabled, the attention module skips the `slice_token.mean(dim=1, keepdim=True)` collapse and instead passes per-head slice tokens directly to `to_k` and `to_v`. The `nn.Linear` projections and downstream einsum shapes handle this naturally since `to_k`/`to_v` operate on the last dim (d_head) which is unchanged.

No parameter count change. No training instability observed — loss progressed smoothly without spikes.

### Metrics (EMA best model)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| p_in | 12.807 | 12.566 | 12.687 | 11.979 | +5.9% ❌ |
| p_oodc | 9.804 | 8.240 | 9.022 | 7.643 | +18.0% ❌ |
| p_tan | 29.112 | 27.952 | 28.532 | 28.341 | +0.7% |
| p_re | 7.671 | 6.746 | 7.209 | 6.300 | +14.4% ❌ |

### Details

- **Seed 42**: W&B `prl91gc2`, best epoch 147/147, val_loss=0.4163
- **Seed 73**: W&B `wbjzazp4`, best epoch 146/146, val_loss=0.3834
- **Epochs completed**: 147 (s42), 146 (s73) — wall-clock limit at 180 min
- **VRAM**: ~50-71 GB (similar to baseline — no meaningful VRAM increase)
- **Training stability**: No spikes or instability. Loss progressed smoothly.

### Command used

```bash
cd cfd_tandemfoil && python train.py --agent askeladd \
  --wandb_name "askeladd/per-head-kv-s{42,73}" \
  --wandb_group "askeladd/per-head-kv-slice" \
  --seed {42,73} \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --per_head_kv
```

### What happened

**Clear negative result — per-head K/V significantly hurts all metrics.**

The shared-mean K/V bottleneck was not a limitation but rather an essential implicit regularization. Every head being forced to query the same mean-pooled K/V bank acts as a communication mechanism that prevents heads from overfitting to narrow per-head patterns.

Key observations:
- **p_oodc** regressed by +18.0% — the largest regression. This suggests per-head K/V enables heads to overfit to their own idiosyncratic patterns rather than learning generalizable features. OOD-condition generalization requires shared understanding across heads.
- **p_re** regressed by +14.4% — same story for OOD-Reynolds generalization.
- **p_in** regressed by +5.9% — even in-distribution performance worsened, suggesting the model is not just overfitting to ID data but actually learning worse representations overall.
- **p_tan** only regressed +0.7% — tandem transfer was least affected, interestingly. The hypothesis that per-head KV would help inter-foil communication may have some merit but is overwhelmed by the overall degradation.
- **Seed 42 much worse than seed 73** (val_loss 0.416 vs 0.383) — the increased variance between seeds suggests the per-head formulation is less stable/robust, even if individual training runs don't spike.

The shared mean acts as an information bottleneck that forces heads to encode complementary information. When each head has its own K/V bank, heads can develop redundant or conflicting representations. This is consistent with literature on bottleneck attention (e.g., Set Transformer's Induced Set Attention Block uses shared inducing points for exactly this reason).

### Suggested follow-ups

1. **Partial per-head KV**: Use per-head KV for the last block only (where head specialization matters most for output) while keeping shared KV in earlier blocks for feature extraction.
2. **Mixed KV**: `K = self.to_k(slice_token) + self.to_k_shared(slice_token.mean(dim=1, keepdim=True))` — per-head + shared, so each head has its own KV but also has access to the global mean.
3. **Per-head KV with explicit regularization**: Add a head coherence loss that penalizes K/V divergence across heads, replacing the implicit regularization with an explicit one.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2265 because ADVISOR: Closing — all 4 metrics regressed, especially p_oodc (+18.0%) and p_re (+14.4%).

**Root cause:** The shared-mean K/V projection in Physics_Attention_Irregular_Mesh acts as essential regularization. Splitting into per-head projections removes this implicit weight-sharing prior, leading to overfitting. The massive p_oodc regression (+18.0%) confirms this — per-head K/V destroys OOD generalization.

**Additional factor:** T_max=160 was used instead of baseline T_max=150, which may have contributed to the regression (T_max sweep already confirmed 150 is optimal).

**Takeaway:** The shared K/V projection is load-bearing regularization in the physics attention mechanism. Per-head specialization at the attention level requires a different approach — perhaps per-head queries only with shared K/V, but the gains would likely be marginal. Added to DO NOT REVISIT..
